File: internal/service/import_svc.go
package service

import (
	"context"
	"fmt"
	"io"
	"log/slog"
	"time"

	"github.com/zajca/zfaktury/internal/domain"
)

// ImportService orchestrates expense creation from an uploaded document.
type ImportService struct {
	expenses  *ExpenseService
	documents *DocumentService
	ocr       *OCRService // may be nil
}

// NewImportService creates a new ImportService.
func NewImportService(expenses *ExpenseService, documents *DocumentService, ocr *OCRService) *ImportService {
	return &ImportService{
		expenses:  expenses,
		documents: documents,
		ocr:       ocr,
	}
}

// ImportFromDocument creates a skeleton expense, uploads the document, and
// optionally runs OCR. Returns the combined result.
func (s *ImportService) ImportFromDocument(ctx context.Context, filename, contentType string, data io.Reader) (*domain.ImportResult, error) {
	// Create a skeleton expense with minimal placeholder data.
	expense := &domain.Expense{
		Description:     filename,
		Amount:          100, // 1 CZK minimum (in halere)
		IssueDate:       time.Now().Truncate(24 * time.Hour),
		CurrencyCode:    domain.CurrencyCZK,
		BusinessPercent: 100,
		PaymentMethod:   "bank_transfer",
	}

	if err := s.expenses.Create(ctx, expense); err != nil {
		return nil, fmt.Errorf("creating skeleton expense: %w", err)
	}

	// Upload the document linked to the new expense.
	doc, err := s.documents.Upload(ctx, expense.ID, filename, contentType, data)
	if err != nil {
		// Rollback: delete the skeleton expense.
		if delErr := s.expenses.Delete(ctx, expense.ID); delErr != nil {
			slog.Error("failed to rollback skeleton expense after upload failure", "error", delErr, "expense_id", expense.ID)
		}
		return nil, fmt.Errorf("uploading document: %w", err)
	}

	result := &domain.ImportResult{
		Expense:  *expense,
		Document: *doc,
	}

	// Run OCR if configured -- non-fatal on failure.
	if s.ocr != nil {
		ocrResult, ocrErr := s.ocr.ProcessDocument(ctx, doc.ID)
		if ocrErr != nil {
			slog.Warn("OCR processing failed during import, continuing without OCR", "error", ocrErr, "document_id", doc.ID)
		} else {
			result.OCR = ocrResult
		}
	}

	return result, nil
}

File: rust/crates/zfaktury-core/src/service/import_svc.rs
use std::sync::Arc;

use zfaktury_domain::{Amount, CURRENCY_CZK, DomainError, Expense};

use super::document_svc::DocumentService;
use super::expense_svc::ExpenseService;
use super::ocr_svc::OCRService;

/// Service for importing expenses from uploaded documents.
pub struct ImportService {
    expenses: Arc<ExpenseService>,
    documents: Arc<DocumentService>,
    ocr: Option<Arc<OCRService>>,
}

/// Result of an import operation.
pub struct ImportResult {
    pub expense: Expense,
}

impl ImportService {
    pub fn new(
        expenses: Arc<ExpenseService>,
        documents: Arc<DocumentService>,
        ocr: Option<Arc<OCRService>>,
    ) -> Self {
        Self {
            expenses,
            documents,
            ocr,
        }
    }

    /// Creates a skeleton expense for an imported document.
    /// The actual file upload is handled by the caller/handler layer.
    pub fn create_skeleton_expense(&self, filename: &str) -> Result<Expense, DomainError> {
        let now = chrono::Local::now().naive_local();
        let today = chrono::Local::now().date_naive();
        let mut expense = Expense {
            id: 0,
            vendor_id: None,
            vendor: None,
            expense_number: String::new(),
            category: String::new(),
            description: filename.to_string(),
            issue_date: today,
            amount: Amount::new(1, 0), // 1 CZK minimum placeholder
            currency_code: CURRENCY_CZK.to_string(),
            exchange_rate: Amount::ZERO,
            vat_rate_percent: 0,
            vat_amount: Amount::ZERO,
            is_tax_deductible: false,
            business_percent: 100,
            payment_method: "bank_transfer".to_string(),
            document_path: String::new(),
            notes: String::new(),
            tax_reviewed_at: None,
            items: Vec::new(),
            created_at: now,
            updated_at: now,
            deleted_at: None,
        };

        self.expenses.create(&mut expense)?;
        Ok(expense)
    }
}
